Software/Drivers/Src/Contactors.c

```diff
@@ -0,0 +1,134 @@
+        .sense_pin  = { HV_NEGATIVE_SENSE_PORT,  HV_NEGATIVE_SENSE_PIN_NUM  }
+    };
+    
+    contactors[Array_contactor] = (contactor_t){
+        .control_pin = { ARRAY_CONTROL_PORT, ARRAY_CONTROL_PIN_NUM },
+        .sense_pin  = { ARRAY_SENSE_PORT,  ARRAY_SENSE_PIN_NUM  }
+    };
+
+    contactors[Array_Pre_contactor] = (contactor_t){
+        .control_pin = { ARRAY_PRE_CONTROL_PORT, ARRAY_PRE_CONTROL_PIN_NUM },
+        .sense_pin  = { ARRAY_PRE_SENSE_PORT,  ARRAY_PRE_SENSE_PIN_NUM  }
+    };
+
+    // loop to intialize contactor timers 
+    for (uint8_t contactor_num = 0; contactor_num < NUM_CONTACTORS; contactor_num++) {
+        // making timers and putting them into structs
+        contactors[contactor_num].senseTimer = xTimerCreateStatic(
+            CONTACTOR_NAMES[contactor_num],                     /* Name of the timer */
+            CONTACTOR_SENSE_DELAY,                              /* Timer period in ticks */
+            pdFALSE,                                            /* Don't auto-reload */
+            (void*)contactor_num,                               /* Timer ID */
+            vContactorCallback,                                 /* Callback function */
+            &(contactors[contactor_num].senseTimerBuffer)       /* Buffer to hold timer data */
+        );
+    }
+
+    // INITIALIZE MUTEX
+    contactorsMutex = xSemaphoreCreateMutexStatic(&contactorsMutexBuffer);
+}
+
+bool Contactors_Get(contactor_enum_t contactor_num) {
+
+    contactor_t* contactor = &contactors[contactor_num];
+    return HAL_GPIO_ReadPin(contactor->sense_pin.port, contactor->sense_pin.pin_num);
+
+}
+
+void vContactorCallback( TimerHandle_t senseTimer ) {
+
+    contactor_enum_t contactor_num = (contactor_enum_t)pvTimerGetTimerID(senseTimer);
+    contactor_t* contactor = &contactors[contactor_num];
+
+    // take mutex for a short time just to confirm check is accurate
+    if (xSemaphoreTake(contactorsMutex, CALLBACK_BLOCKING_TIME) == pdFALSE) {
+        // Someone is using the contactor sephmore for a lot longer than they should. call fault handler.
+    }
+    if (contactor->state != Contactors_Get(contactor_num)) {
+        // This means the contactor actual state is not what the code thinks it is. call fault handler.
+    }
+
+    // return mutex
+    if (xSemaphoreGive(contactorsMutex) == pdFALSE) {
+        // this really shouldn't happen. call fault handler.
+    }
+ 
+}
+
+ErrorStatus toggle_contactor(contactor_enum_t contactor_num, bool state, bool blocking, bool emergency) {
+    
+    contactor_t* contactor = &contactors[contactor_num];
+
+    // if its emergency, dont bother with semaphore
+    if (!emergency && xSemaphoreTake(contactorsMutex, blocking ? portMAX_DELAY : 0) == pdFALSE) {
+        return ERROR;
+    };
+
+    // critical section:
+    HAL_GPIO_WritePin(contactor->control_pin.port, contactor->control_pin.pin_num, state);
+    contactor->state = state;
+    
```

Comment:
@Lakshay983 can probably comment better, but if this is truly an emergency critical section, would it be worth it to disable interrupts around this? I'm not totally sure. It's probably only 10-12 lines of assembly code, so the chances of the RTOS context-switching us in between these two lines is low.